Chart filtering

j6ll opened the pull request:
## Proposed Changes

* Adds date filtering to the Blaze stats charts
* Adds loading spinner
* Adds hourly/daily resolutions

![Screenshot 2024-11-28 at 19 09 14](https://github.com/user-attachments/assets/ff096505-c2be-47f0-8855-2392c27b43c5)

## Testing Instructions

Here are some scenarios that can be tested, some of them require specific data that could be hard to replicate so you can check the code for these too,  I have also provided a demo video to show them in use.

- [ ] Check the different range options work
- [ ] Yesterday / Today disappear if campaign has finished
- [ ] Last X days will be relative to today in an active campaign
- [ ] Last X days will be relative to the end date of a finished campaign
- [ ] Last X days won't show on a campaign that is less than X
- [ ] Today and Yesterday will fetch hourly data


Her is a demo of the above scenarios

https://github.com/user-attachments/assets/e8593dbe-b768-40cf-95df-2cae249bf178


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

j6ll committed:
Chart filtering:
- Adds new filtering options for the chart
- Shows data as hourly
- Adds loading spinner

j6ll committed:
Chart filtering:
- Adds new filtering options for the chart
- Shows data as hourly
- Adds loading spinner

j6ll commented on client/my-sites/promote-post-i2/components/campaign-item-details/index.tsx:
```diff
@@ -163,10 +180,64 @@ export default function CampaignItemDetails( props: Props ) {
 		conversion_last_currency_found,
 	} = campaign_stats || {};
 
+	const getChartStartDate = ( dateRange: ChartSourceDateRanges ) => {
```
I'd like to refactor these changes, so that they are in their own file, but in the interest of time, I'm going to follow up with that.

timur987 reviewed (commented):
Thanks, @j6ll!

Event though I found this changes great and working as advertised I want them to be even better. Here's the patch which I think would be interesting for you.

[patch.txt](https://github.com/user-attachments/files/17957721/patch.txt)

Shortly the list of proposed changes are the following:

- `client/data/promote-post/use-campaign-chart-stats-query.ts`
  - Add endDate param to show "Yesterday" correctly (see the image below),
  - use user's timezone instead of UTC

Before I got, for yesterday, start not at 12:00 AM, but instead the same using UTC;
Another thing is having today's stat on graph.
![Screenshot 2024-11-29 at 13 25 49](https://github.com/user-attachments/assets/6fd383f1-b0af-49fe-af26-e0bf7feed218)

It should be fixed and looks like this:
![Screenshot 2024-11-29 at 13 21 19](https://github.com/user-attachments/assets/160f9d79-be5d-42d9-8153-9b01601c4711)

- `client/my-sites/promote-post-i2/components/campaign-item-details/index.tsx`
 - I do think we should not request the whole campaign stats by default and show last 7 days instead (maybe I'm wrong though),
 - your code to getting an effective end date is useful and can be extracted to a separate function,
 - in order to fix "yesterday's" issue, we can send endDate to the query.


That's it.
Please take it with a pinch of salt 🧂 .

Leaving it as a comment, but will be ready to approve if you consider these thoughts and not appropriate. Anyway, please let me know what do you think.

j6ll committed:
- Set default to 7 days
- Extract effective end date calculation
- Set to users timezone rather than UTC

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/17041919

Some locales (Hebrew, Japanese) have been temporarily machine-translated due to translator availability. All other translations are usually ready within a few days. Untranslated and machine-translated strings will be sent for translation next Monday and are expected to be completed by the following Friday.

Thank you @j6ll for including a screenshot in the description! This is really helpful for our translators.
